### ISSUE TITLE
What is valid syntax inside a module changes after executing `:int.i/1`

### ISSUE THREAD

**Issue by jdicioccio** [elixir-lang/elixir#4857]
### Environment

* Elixir version (elixir -v):
```
Erlang/OTP 18 [erts-7.3] [source] [64-bit] [smp:8:8] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Elixir 1.3.0
```

* Operating system: OS X 10.11.5

### Current behavior

After using `:int.i/1` on a loaded module that loops into itself, anonymous functions in that module begin to throw errors.

See https://github.com/jdicioccio/ex_repro for simple repro

### Expected behavior

It would be expected that syntax would be the same after executing `:int.i/1`

**Comment by josevalim** [comment]
@jdicioccio that's unfortunately how anonymous functions and code loading works in Erlang. Anonymous function names are generated somewhat dynamically so when you interpret the module, it loads a new module version into memory, and the anonymous functions get new names and the old ones can no longer be found. We usually properly update the code in production (for example, like in a GenServer code_change callback) so we do not run into such issues.

**Comment by cjhowe** [comment]
Your answer doesn't provide any solution to the problem. Presenting another problem does not resolve this issue. Please consider the detriment this type of mistake has on the community. We don't want people being erroneously convinced that they have made a mistake or that no solution exists.

If he started an IEx shell with the interpreter enabled at the start, so that the initial state is correct, I believe it would solve the problem. Maybe he can just enable the interpreter in his code before it starts the process?

**Comment by josevalim** [comment]
> Presenting another problem does not resolve this issue.

The previous comment explained *why* he is encountering the behaviour he described. It is a VM limitation.

I agree, however, I did not provide a solution to the problem. But couldn't you just have asked for one? The fact my answer was incomplete was not on purpose or without consideration.

That said, if you upgrade/interpret a code while a process holds references to its anonymous function, those references will *break*. The solution is to interpret the code before starting the process (or start a new one after).

**Comment by josevalim** [comment]
@cjhowe and if in turn I read too much into your comment, I apologize. ❤️

**Comment by cjhowe** [comment]
Unfortunately in this case, unintentional or not, it could have had an negative impact on the growth of the community. I think it's appropriate to be critical of mistakes that cause harm to the community, even if it's unintentional. If necessary, maybe having two people check each issue before closing it would help increase the quality of issue triage?

**Comment by antipax** [comment]
@cjhowe While I agree that providing a solution would have been optimal, in my humble opinion I believe you're being somewhat hyperbolic by calling this a harmful mistake. 

The issue tracker is here for the purpose of tracking bugs/features for the community, and that @josevalim took the time at all to explain why this wasn't a bug, rather than just closing the issue as "won't fix" was above and beyond what you would get from many other project maintainers. The Elixir IRC/Slack/Forum is where support requests belong, not here on the issue tracker.

That being said, it is always possible to do better, and in the future when closing issues it might be a good idea to point users in the direction of those resources in order to find a solution to their non-bug issues. Additionally, as a member of the community, even though I disagree with you, I would like to thank you for your concern, as it benefits all of us. ❤️

**Comment by josevalim** [comment]
> Unfortunately in this case, unintentional or not, it could have had an negative impact on the growth of the community. I think it's appropriate to be critical of mistakes that cause harm to the community, even if it's unintentional.

Good points, agreed.

> If necessary, maybe having two people check each issue before closing it would help increase the quality of issue triage?

I believe we have different perceptions of how the issues tracker is used. This issue was triaged properly.  Closing the issue does not imply end of the discussion, "closed" reflects there is nothing for contributors to work on. Sometimes we may close improperly but we always reopen as soon as *anyone* asks. The conversation is always on-going, as it is happening now.
